### PR TITLE
programmatic client generation for MP Rest Client, fixes #1317

### LINKED ISSUE #1317: Programmatic rest client generation fails in native mode
To reproduce, you may use the `CountriesResourceTest` in https://github.com/michalszynkiewicz/quarkus-quickstarts/tree/builder-native-failure

Error log:
https://gist.github.com/michalszynkiewicz/174fbfabfb1bb14ccd36f3a8c2d533c5

The cause is that we only register interfaces annotated with `@RegisterRestClient` here:
https://github.com/quarkusio/quarkus/blob/13f7c14bc28714c866e204e7fc3e7acafdd13a56/extensions/smallrye-rest-client/deployment/src/main/java/io/quarkus/smallrye/restclient/deployment/SmallRyeRestClientProcessor.java#L171-L175

### PR DESCRIPTION
**Pull request by michalszynkiewicz**

### CODE REVIEW

**Comment by geoand** [comment]
@michalszynkiewicz tests are failing

**Comment by michalszynkiewicz** [comment]
@geoand the tests were fine, a couple unused imports sneaked in

**Comment by geoand** [comment]
@michalszynkiewicz cool :)

**Comment by geoand** [comment]
@michalszynkiewicz I assume that when this is merged the [quickstart](https://github.com/quarkusio/quarkus-quickstarts/blob/master/rest-client/src/main/java/org/acme/restclient/CountriesService.java)  will need to be updated as well?

File: integration-tests/rest-client/pom.xml [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+  <parent>
+    <artifactId>quarkus-integration-tests-parent</artifactId>
+    <groupId>io.quarkus</groupId>
+    <version>999-SNAPSHOT</version>
+    <relativePath>../</relativePath>
+  </parent>
+  <modelVersion>4.0.0</modelVersion>
+
+  <artifactId>quarkus-integration-test-rest-client</artifactId>
```
Comment (gsmet) [inline integration-tests/rest-client/pom.xml]: I think the rest client tests are integrated in the main integration tests. It should stay that way.

**Review by gsmet — CHANGES_REQUESTED** [review]
Tests should be integrated with the others.

File: integration-tests/rest-client/pom.xml (hunk 1 above)
Comment (michalszynkiewicz) [reply integration-tests/rest-client/pom.xml]: I missed that, thanks!

**Comment by michalszynkiewicz** [comment]
@geoand what's currently in the quickstarts will work. They don't cover programmatic creation of the client, I don't know if it's something that should be/is worth adding to the quickstarts

**Comment by geoand** [comment]
Got it, thanks! Maybe then we could add a section about the improved support you added in the doc about the rest-client?

File: integration-tests/rest-client/pom.xml (hunk 1 above)
Comment (michalszynkiewicz) [reply integration-tests/rest-client/pom.xml]: done, I removed my tests and modified the existing ones to use a client interface without `@RegisterRestClient`.
In the process I removed the test for not marking the interface that has an implementation as eligible for proxying in the native image but I think it's for the good ;)

**Comment by michalszynkiewicz** [comment]
the native build failed on a timeout to download smallrye metrics artifact from the jboss nexus.
I retriggered it.

File: integration-tests/main/src/main/java/io/quarkus/example/rest/ProgrammaticRestInterface.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+
+/**
+ * A version of {@link RestInterface} that doesn't have {@link org.eclipse.microprofile.rest.client.inject.RegisterRestClient}
+ * and can be used only programmatically, i.e. with the builder.
+ *
+ * @author Michal Szynkiewicz, michal.l.szynkiewicz@gmail.com
+ *         <br>
+ *         Date: 3/15/19
```
Comment (gsmet) [inline integration-tests/main/src/main/java/io/quarkus/example/rest/ProgrammaticRestInterface.java]: I would drop that. We don't really use that, we have the git history.

File: docs/src/main/asciidoc/rest-client-guide.adoc [hunk 3]
```diff
@@ -268,3 +269,7 @@ You should see a JSON object containing some basic information about Greece.
 
 As usual, the application can be packaged using `mvn clean package` and executed using the `-runner.jar` file.
 You can also generate the native executable with `mvn clean package -Pnative`.
+
+== Further reading
+
+link:https://download.eclipse.org/microprofile/microprofile-rest-client-1.2.1/microprofile-rest-client-1.2.1.html[MicroProfile Rest Client specification]
```
Comment (gsmet) [inline docs/src/main/asciidoc/rest-client-guide.adoc]: Very minor nitpicking: even if there is only one element, I would use a list to have a better rendering.

**Review by gsmet — CHANGES_REQUESTED** [review]
@michalszynkiewicz the implementation looks good.

I added 2 very minor comments, could you adjust and I'll merge?

**Review by gsmet — APPROVED** [review]
Let's wait for CI to go green.

**Pull request merged**